mm2src/coins/nft/trading_proto_v2/mod.rs

```diff
@@ -0,0 +1,565 @@
+use crate::coin_errors::{ValidatePaymentError, ValidatePaymentResult};
+use ethabi::{Contract, Token};
+use ethcore_transaction::{Action, UnverifiedTransaction};
+use ethereum_types::{Address, U256};
+use futures::compat::Future01CompatExt;
+use mm2_err_handle::prelude::{MapToMmResult, MmError, MmResult};
+use mm2_number::BigDecimal;
+use std::convert::TryInto;
+use web3::types::{Transaction as Web3Tx, TransactionId};
+
+pub(crate) mod errors;
+use errors::{Erc721FunctionError, HtlcParamsError, PaymentStatusErr};
+pub(crate) mod structs;
+use structs::{ExpectedHtlcParams, StateType, ValidationParams};
+
+use super::ContractType;
+use crate::eth::{addr_from_raw_pubkey, decode_contract_call, EthCoin, EthCoinType, MakerPaymentStateV2, SignedEthTx,
+                 TryToAddress, ERC1155_CONTRACT, ERC721_CONTRACT, ETH_GAS, NFT_SWAP_CONTRACT};
+use crate::nft::trading_proto_v2::errors::PrepareTxDataError;
+use crate::{NftAssocTypes, RefundPaymentArgs, SendNftMakerPaymentArgs, SpendNftMakerPaymentArgs, TransactionErr,
+            ValidateNftMakerPaymentArgs};
+
+impl EthCoin {
+    pub(crate) async fn send_nft_maker_payment_v2_impl(
+        &self,
+        args: SendNftMakerPaymentArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let contract_type = try_tx_s!(self.parse_contract_type(args.nft_swap_info.contract_type));
+        try_tx_s!(self.validate_payment_args(
+            args.taker_secret_hash,
+            args.maker_secret_hash,
+            &args.amount,
+            &contract_type,
+        ));
```

Comment:
If we decide we need this validation for any type of NFT implementation, it should be part of the state machine also, again, just some considerations for the future so that we have the NFT swap state machine implementation as generic as it can be and make new NFT protocol implementations easier. I guess we will need to visualize the state machine diagram to decide how to break up some of the implementations in this file to generic trait methods if needed.

Reply:
will make a note about this in nft issue

Reply:
here https://github.com/KomodoPlatform/komodo-defi-framework/issues/900#issuecomment-2039526750